File: BlazorApp1/llm_client.py
import requests
import os
from openai import OpenAI

class LLMClient:
    """Unified client for both cloud AI (OpenAI) and local AI (Ollama)"""
    
    def __init__(self, config):
        self.config = config
        self.provider = config.ai_provider
        
        if self.provider == "openai":
            self.client = OpenAI(api_key=config.openai_api_key)
            self.model = config.openai_model
        elif self.provider == "ollama":
            self.base_url = config.ollama_base_url
            self.model = config.ollama_model
        else:
            raise ValueError(f"Unknown AI provider: {self.provider}")
    
    def get_completion(self, system_prompt, user_prompt):
        """Get completion from LLM"""
        if self.provider == "openai":
            return self._get_openai_completion(system_prompt, user_prompt)
        elif self.provider == "ollama":
            return self._get_ollama_completion(system_prompt, user_prompt)
    
    def _get_openai_completion(self, system_prompt, user_prompt):
        """Get completion from OpenAI API"""
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt}
                ],
                temperature=0.7,
                max_tokens=2000
            )
            return response.choices[0].message.content
        except Exception as e:
            print(f"OpenAI API error: {e}")
            raise
    
    def _get_ollama_completion(self, system_prompt, user_prompt):
        """Get completion from Ollama API"""
        try:
            url = f"{self.base_url}/api/chat"
            payload = {
                "model": self.model,
                "messages": [
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt}
                ],
                "stream": False,
                "options": {
                    "temperature": 0.7,
                    "num_predict": 2000
                }
            }
            
            response = requests.post(url, json=payload, timeout=120)
            response.raise_for_status()
            
            result = response.json()
            return result["message"]["content"]
            
        except requests.exceptions.RequestException as e:
            print(f"Ollama API error: {e}")
            raise

File: BlazorApp1/BlazorApp1/Services/LlmClient.cs
using System.Net.Http.Json;
using System.Text.Json.Serialization;

namespace BlazorApp1.Services;

public class LlmClient
{
    private readonly AppConfig _config;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly ILogger<LlmClient> _logger;

    public LlmClient(AppConfig config, IHttpClientFactory httpClientFactory, ILogger<LlmClient> logger)
    {
        _config = config;
        _httpClientFactory = httpClientFactory;
        _logger = logger;
    }

    public async Task<string> GetCompletionAsync(string systemPrompt, string userPrompt)
    {
        return _config.AiProvider.ToLowerInvariant() switch
        {
            "openai" => await GetOpenAiCompletionAsync(systemPrompt, userPrompt),
            "ollama" => await GetOllamaCompletionAsync(systemPrompt, userPrompt),
            _ => throw new InvalidOperationException($"Unknown AI provider: {_config.AiProvider}")
        };
    }

    private async Task<string> GetOpenAiCompletionAsync(string systemPrompt, string userPrompt)
    {
        var client = _httpClientFactory.CreateClient("OpenAI");
        client.DefaultRequestHeaders.Authorization =
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _config.OpenAiApiKey);

        var request = new OpenAiRequest
        {
            Model = _config.OpenAiModel,
            Messages =
            [
                new ChatMessage { Role = "system", Content = systemPrompt },
                new ChatMessage { Role = "user", Content = userPrompt }
            ],
            Temperature = 0.7,
            MaxTokens = 2000
        };

        var response = await client.PostAsJsonAsync("https://api.openai.com/v1/chat/completions", request);
        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadFromJsonAsync<OpenAiResponse>();
        return result?.Choices?.FirstOrDefault()?.Message?.Content
               ?? throw new InvalidOperationException("Empty response from OpenAI.");
    }

    private async Task<string> GetOllamaCompletionAsync(string systemPrompt, string userPrompt)
    {
        var client = _httpClientFactory.CreateClient("Ollama");

        var request = new OllamaRequest
        {
            Model = _config.OllamaModel,
            Messages =
            [
                new ChatMessage { Role = "system", Content = systemPrompt },
                new ChatMessage { Role = "user", Content = userPrompt }
            ],
            Stream = false,
            Options = new OllamaOptions { Temperature = 0.7, NumPredict = 2000 }
        };

        var response = await client.PostAsJsonAsync($"{_config.OllamaBaseUrl}/api/chat", request);
        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadFromJsonAsync<OllamaResponse>();
        return result?.Message?.Content
               ?? throw new InvalidOperationException("Empty response from Ollama.");
    }

    #region DTOs

    private sealed class ChatMessage
    {
        [JsonPropertyName("role")] public string Role { get; set; } = "";
        [JsonPropertyName("content")] public string Content { get; set; } = "";
    }

    private sealed class OpenAiRequest
    {
        [JsonPropertyName("model")] public string Model { get; set; } = "";
        [JsonPropertyName("messages")] public List<ChatMessage> Messages { get; set; } = [];
        [JsonPropertyName("temperature")] public double Temperature { get; set; }
        [JsonPropertyName("max_tokens")] public int MaxTokens { get; set; }
    }

    private sealed class OpenAiResponse
    {
        [JsonPropertyName("choices")] public List<OpenAiChoice>? Choices { get; set; }
    }

    private sealed class OpenAiChoice
    {
        [JsonPropertyName("message")] public ChatMessage? Message { get; set; }
    }

    private sealed class OllamaRequest
    {
        [JsonPropertyName("model")] public string Model { get; set; } = "";
        [JsonPropertyName("messages")] public List<ChatMessage> Messages { get; set; } = [];
        [JsonPropertyName("stream")] public bool Stream { get; set; }
        [JsonPropertyName("options")] public OllamaOptions Options { get; set; } = new();
    }

    private sealed class OllamaOptions
    {
        [JsonPropertyName("temperature")] public double Temperature { get; set; }
        [JsonPropertyName("num_predict")] public int NumPredict { get; set; }
    }

    private sealed class OllamaResponse
    {
        [JsonPropertyName("message")] public ChatMessage? Message { get; set; }
    }

    #endregion
}
